### PR TITLE
Feature/issue 142 feature support ipv4 and ipv6 data type

### LINKED ISSUE #142: Support ipv4 and ipv6 data type
for web analytics, IP address is a very common data type. refer to https://clickhouse.com/blog/geolocating-ips-in-clickhouse-and-grafana.

- https://clickhouse.com/docs/en/sql-reference/data-types/ipv4
- https://clickhouse.com/docs/en/sql-reference/data-types/ipv6

it is better to have such type support in proton

### PR DESCRIPTION
**Pull request by lizhou1111**
this close #142 
support ipv4 and ipv6 data type

- [x] https://github.com/ClickHouse/ClickHouse/pull/49759
- [x] https://github.com/ClickHouse/ClickHouse/pull/43221
- [x] https://github.com/ClickHouse/ClickHouse/pull/48556
- [x] https://github.com/ClickHouse/ClickHouse/pull/45425
- [x] https://github.com/ClickHouse/ClickHouse/pull/47208
- [x] https://github.com/ClickHouse/ClickHouse/pull/45043
- [x] https://github.com/ClickHouse/ClickHouse/pull/48486
- [x] https://github.com/ClickHouse/ClickHouse/pull/45799
- [x] https://github.com/ClickHouse/ClickHouse/pull/49933
- [x] https://github.com/ClickHouse/ClickHouse/pull/48486

### CODE REVIEW

**Comment by yokofly** [comment]
This is a pull request block that requires signing the CLA. You can squash all the commits and rebase to rewrite the author name.  Also, plz give the all related PR link in comment.



@lizhou1111  can we stop the CI action?

File: src/Interpreters/convertFieldToType.cpp [hunk 1]
```diff
@@ -244,12 +244,37 @@ Field convertFieldToTypeImpl(const Field & src, const IDataType & type, const ID
...
+        /// Already in needed type.
+        return src;
+    }
+    else if (which_type.isIPv6())
+    {
+        /// Already in needed type.
+        if (src.getType() == Field::Types::IPv6)
+            return src;
+        /// Treat FixedString(16) as a binary representation of IPv6
+        if (which_from_type.isFixedString() && assert_cast<const DataTypeFixedString *>(from_type_hint)->getN() == IPV6_BINARY_LENGTH)
```
Comment (yl-lisen) [inline src/Interpreters/convertFieldToType.cpp]: looks different with community latest code

File: src/IO/WriteHelpers.cpp [hunk 2]
```diff
@@ -34,6 +36,78 @@ void formatUUID(std::reverse_iterator<const UInt8 *> src16, UInt8 * dst36)
...
+    buf.write(byte_str[idx].first, byte_str[idx].second);
+    buf.write('.');
+    idx = (ip >> 16) & 0xFF;
+    buf.write(byte_str[idx].first, byte_str[idx].second);
+    buf.write('.');
+    idx = (ip >> 8) & 0xFF;
+    buf.write(byte_str[idx].first, byte_str[idx].second);
+    buf.write('.');
+    idx = ip & 0xFF;
+    buf.write(byte_str[idx].first, byte_str[idx].second);
```
Comment (yl-lisen) [inline src/IO/WriteHelpers.cpp]: here looks different with latest community code

File: src/Functions/isIPAddressContainedIn.cpp [hunk 3]
```diff
@@ -1,6 +1,7 @@
 #include <Columns/ColumnConst.h>
 #include <Columns/ColumnString.h>
 #include <Columns/ColumnsNumber.h>
+#include "Common/Exception.h"
```
Comment (yl-lisen) [inline src/Functions/isIPAddressContainedIn.cpp:4]: here looks different with latest community code

File: src/Functions/FunctionsCodingIP.cpp [hunk 4]
```diff
@@ -1,3 +1,6 @@
+#include "Core/Types.h"
+#include "base/IPv4andIPv6.h"
+#include "base/types.h"
```
Comment (yl-lisen) [inline src/Functions/FunctionsCodingIP.cpp]: here looks different with latest community code

File: src/Functions/FunctionsCodingIP.cpp [hunk 5]
```diff
@@ -7,21 +10,22 @@
...
+#include <DataTypes/DataTypeIPv4andIPv6.h>
 #include <DataTypes/DataTypeString.h>
 #include <DataTypes/DataTypeTuple.h>
 #include <DataTypes/DataTypesNumber.h>
 #include <Functions/FunctionFactory.h>
 #include <Functions/FunctionHelpers.h>
 #include <Functions/IFunction.h>
-#include <Interpreters/Context.h>
 #include <IO/WriteHelpers.h>
+#include <Interpreters/Context.h>
```
Comment (yl-lisen) [inline src/Functions/FunctionsCodingIP.cpp]: here looks different with latest community code

File: src/Functions/FunctionsCodingIP.cpp [hunk 6]
```diff
@@ -7,21 +10,22 @@
 #include <Columns/ColumnArray.h>
 #include <Columns/ColumnConst.h>
 #include <Columns/ColumnFixedString.h>
+#include <Columns/ColumnNullable.h>
```
Comment (yl-lisen) [inline src/Functions/FunctionsCodingIP.cpp]: here looks different with latest community code

File: src/Common/ErrorCodes.cpp [hunk 7]
```diff
@@ -624,6 +624,8 @@
     M(666, CANNOT_USE_CACHE) \
     M(677, THREAD_WAS_CANCELED) \
     M(678, IO_URING_INIT_FAILED) \
+    M(669, CANNOT_PARSE_IPV4) \
+    M(670, CANNOT_PARSE_IPV6) \
```
Comment (yl-lisen) [inline src/Common/ErrorCodes.cpp]: here looks different with latest community code.

File: src/AggregateFunctions/AggregateFunctionMinMaxK.cpp [hunk 8]
```diff
@@ -1,13 +1,15 @@
 #include "AggregateFunctionMinMaxK.h"
 
 #include "AggregateFunctionFactory.h"
+#include "DataTypes/DataTypeIPv4andIPv6.h"
```
Comment (yl-lisen) [inline src/AggregateFunctions/AggregateFunctionMinMaxK.cpp]: remove duplicated header

**Review by yl-lisen — APPROVED** [review]

**Pull request merged**